contracts/MechManager.sol

```diff
@@ -0,0 +1,279 @@
+    mapping(address => uint256) public mapRequestCounts;
+    // Map of delivery counts for corresponding requester
+    mapping(address => uint256) public mapDeliveryCounts;
+    // Map of delivery counts for agent mech
+    mapping(address => uint256) public mapAgentMechDeliveryCounts;
+    // Map of delivery counts for corresponding mech service multisig
+    mapping(address => uint256) public mapMechServiceDeliveryCounts;
+    // Map of agent mech => its creating factory
+    mapping(address => address) public mapAgentMechFactories;
+
+    /// @dev MechManager implementation constructor.
+    /// @param _serviceRegistry Service registry address.
+    constructor(address _serviceRegistry) {
+        serviceRegistry = _serviceRegistry;
+    }
+
+    /// @dev MechManager initializer.
+    function initialize(uint256 _fee) external{
+        if (owner != address(0)) {
+            revert AlreadyInitialized();
+        }
+
+        owner = msg.sender;
+        fee = _fee;
+    }
+
+    /// @dev Changes the mechManager implementation contract address.
+    /// @param newImplementation New implementation contract address.
+    function changeImplementation(address newImplementation) external {
+        // Check for the ownership
+        if (msg.sender != owner) {
+            revert OwnerOnly(msg.sender, owner);
+        }
+
+        // Check for zero address
+        if (newImplementation == address(0)) {
+            revert ZeroAddress();
+        }
+
+        // Store the mechManager implementation address
+        assembly {
+            sstore(MECH_MANAGER_PROXY, newImplementation)
+        }
+
+        emit ImplementationUpdated(newImplementation);
+    }
+
+    /// @dev Changes contract owner address.
+    /// @param newOwner Address of a new owner.
+    function changeOwner(address newOwner) external virtual {
+        // Check for the ownership
+        if (msg.sender != owner) {
+            revert OwnerOnly(msg.sender, owner);
+        }
+
+        // Check for the zero address
+        if (newOwner == address(0)) {
+            revert ZeroAddress();
+        }
+
+        owner = newOwner;
+        emit OwnerUpdated(newOwner);
+    }
+
+    /// @dev Registers service as a mech.
+    /// @param serviceId Service id.
+    /// @param mechFactory Mech factory address.
+    /// @return mech The created mech instance address.
+    function create(uint256 serviceId, address mechFactory, bytes memory payload) external returns (address mech) {
+        // Check for factory status
+        if (!mapMechFactories[mechFactory]) {
+            revert UnauthorizedAccount(mechFactory);
+        }
+
+        mech = IMechFactory(mechFactory).createMech(address(this), serviceRegistry, serviceId, payload);
+
+        // This should never be the case
+        if (mech == address(0)) {
+            revert ZeroAddress();
+        }
+
+        // Record factory that created agent mech
+        mapAgentMechFactories[mech] = mechFactory;
```

Comment:
Not needed imo.

But where are we recording the mechs that the marketplace has registered on it?

Currently there's no on-chain mapping. That makes discovery very hard.

Reply:
See above